File: zy73185/src/engine/similarity.ts
export function tokenize(text: string): Set<string> {
  const tokens = new Set<string>();
  const cleaned = text.toLowerCase().replace(/[\s，。、；：""''（）\(\)\[\]【】,.;:\-_/\\]/g, " ");
  for (const piece of cleaned.split(/\s+/)) {
    if (piece.length > 0) tokens.add(piece);
  }
  for (let i = 0; i < cleaned.length - 1; i++) {
    const bigram = cleaned.slice(i, i + 2).trim();
    if (bigram.length === 2) tokens.add(bigram);
  }
  return tokens;
}

export function jaccard(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 || b.size === 0) return 0;
  let inter = 0;
  for (const x of a) if (b.has(x)) inter++;
  const union = a.size + b.size - inter;
  return union === 0 ? 0 : inter / union;
}

export function similarityScore(a: string, b: string): number {
  return jaccard(tokenize(a), tokenize(b));
}

File: zy73185/api/engine/similarity.py
import re
from typing import Set

_PUNCT_RE = re.compile(r"[\s，。、；：\"''（）\(\)\[\]【】,.;:\-_/\\]+")


def _tokenize(text: str) -> Set[str]:
    tokens: Set[str] = set()
    cleaned = _PUNCT_RE.sub(" ", text.lower())
    for piece in cleaned.split():
        if piece:
            tokens.add(piece)
    for i in range(len(cleaned) - 1):
        bigram = cleaned[i : i + 2].strip()
        if len(bigram) == 2:
            tokens.add(bigram)
    return tokens


def jaccard(a: Set[str], b: Set[str]) -> float:
    if not a or not b:
        return 0.0
    inter = len(a & b)
    union = len(a | b)
    return inter / union if union else 0.0


def similarity_score(a: str, b: str) -> float:
    return jaccard(_tokenize(a), _tokenize(b))
